packages/devtools_app/test/integration_tests/server_test.dart

```diff
@@ -111,158 +112,178 @@ void main() {
-      await launchDevTools(reuseWindows: true);
-
-      // Ensure we now have a single connected client.
-      final serverResponse =
-          await _waitForClients(requiredConnectionState: true);
-      expect(serverResponse['clients'], hasLength(1));
-      expect(serverResponse['clients'][0]['hasConnection'], isTrue);
-      expect(serverResponse['clients'][0]['vmServiceUri'],
-          equals(appFixture.serviceUri.toString()));
-    }, timeout: const Timeout.factor(10));
-    // The API only works in release mode.
-  }, skip: !testInReleaseMode);
+        // Send a request to launch DevTools in a browser.
+        await launchDevTools(useVmService: useVmService);
+
+        {
+          final serverResponse =
+              await _waitForClients(requiredConnectionState: true);
+          expect(serverResponse['clients'], hasLength(1));
+        }
+
+        // Request again, allowing reuse, and server emits an event saying the
+        // window was reused.
+        final launchResponse = await launchDevTools(
+            useVmService: useVmService, reuseWindows: true);
+        expect(launchResponse['reused'], isTrue);
+
+        // Ensure there's still only one connection (eg. we didn't spawn a new one
+        // we reused the existing one).
+        final serverResponse =
+            await _waitForClients(requiredConnectionState: true);
+        expect(serverResponse['clients'], hasLength(1));
+      }, timeout: const Timeout.factor(10));
+
+      test('Server reuses DevTools instance if not connected to a VM',
+          () async {
+        // Register the VM.
+        await _send('vm.register', {'uri': appFixture.serviceUri.toString()});
+
+        // Send a request to launch DevTools in a browser.
+        await launchDevTools(useVmService: useVmService);
+
+        // Wait for the DevTools to inform server that it's connected.
+        await _waitForClients(requiredConnectionState: true);
+
+        // Terminate the VM.
+        await appFixture.teardown();
+
+        // Ensure the client is marked as disconnected.
+        await _waitForClients(requiredConnectionState: false);
+
+        // Start up a new app.
+        await _startApp();
+        await _send('vm.register', {'uri': appFixture.serviceUri.toString()});
+
+        // Send a new request to launch.
+        await launchDevTools(useVmService: useVmService, reuseWindows: true);
+
+        // Ensure we now have a single connected client.
+        final serverResponse =
+            await _waitForClients(requiredConnectionState: true);
+        expect(serverResponse['clients'], hasLength(1));
+        expect(serverResponse['clients'][0]['hasConnection'], isTrue);
+        expect(serverResponse['clients'][0]['vmServiceUri'],
+            equals(appFixture.serviceUri.toString()));
+      }, timeout: const Timeout.factor(10));
+      // The API only works in release mode.
+    }, skip: !testInReleaseMode);
+  }
 }
 
 Future<Map<String, dynamic>> launchDevTools({
```

Comment:
As a nit, this method is named the same as the one in `lib/src/server.dart`; perhaps rename slightly to disambiguate; make it private, or add `test` somewhere in the name?

Reply:
Done! (06499caa)